File: JavaScript/450. Delete Node in a BST.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} key
 * @return {TreeNode}
 */
var deleteNode = function(root, key) {
    if (root==null){return null;}
    if(root.val<key){
        root.right=deleteNode(root.right,key);
    }
    else if(root.val>key){
        root.left=deleteNode(root.left,key);
    }
    else{
        if(root.left==null && root.right==null){
            return null;
        }
        else if(root.left==null || root.right==null){
            if(root.left!=null){return root.left;}
            return root.right;
        }
        else{
            temp=root.left;
            while(temp.right!=null){
                temp=temp.right;
            }
            root.val=temp.val;
            root.left=deleteNode(root.left, temp.val);
        }
    }
    return root;
};

File: C/450. Delete Node in a BST.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
struct TreeNode* deleteNode(struct TreeNode* root, int key){
        if (root==NULL){return NULL;}
        if(root->val<key){
            root->right=deleteNode(root->right,key);
        }
        else if(root->val>key){
            root->left=deleteNode(root->left,key);
        }
        else{
            if(root->left==NULL && root->right==NULL){
                return NULL;
            }
            else if(root->left==NULL || root->right==NULL){
                if(root->left!=NULL){return root->left;}
                return root->right;
            }
            else{
                struct TreeNode* temp=root->left;
                while(temp->right){
                    temp=temp->right;
                }
                root->val=temp->val;
                root->left=deleteNode(root->left, temp->val);
            }
        }
        return root;
}
